libibumad/umad.h

```diff
@@ -54,6 +54,21 @@ typedef __be16 __attribute__((deprecated)) be16_t;
 typedef __be32 __attribute__((deprecated)) be32_t;
 typedef __be64 __attribute__((deprecated)) be64_t;
 
+typedef union _ib_gid
+{
+	uint8_t	raw[16];
+
+	struct {
+		__be64	prefix;
+		__be64	interface_id;
+	} unicast;
+
+	struct {
+		uint8_t	header[2];
+		uint8_t	raw_group_id[14];
+	} multicast;
+} ib_gid_t;
+
```

Comment:
The ib_mad_notice_attr_t has the PACKED attribute set which does eliminate much of the padding, so the overall size will not change.

However, the ib_gid_t's in that structure are not 8 byte aligned:

```
_Static_assert(__alignof__(ib_gid_t) == 8, "err");
_Static_assert((uintptr_t)(&((ib_mad_notice_attr_t *)0)->data_details.ntc_64_67.gid) % 8 == 0, "err");
_Static_assert((uintptr_t)(&((ib_mad_notice_attr_t *)0)->data_details.ntc_257_258.gid1) % 8 == 0, "err");
_Static_assert((uintptr_t)(&((ib_mad_notice_attr_t *)0)->data_details.ntc_257_258.gid2) % 8 == 0, "err");
_Static_assert((uintptr_t)(&((ib_mad_notice_attr_t *)0)->data_details.ntc_259.gid1) % 8 == 0, "err");
_Static_assert((uintptr_t)(&((ib_mad_notice_attr_t *)0)->data_details.ntc_259.gid2) % 8 == 0, "err");
```

Yields compile failures.

The compiler may generate correct code for

```
ib_gid_t tmp = notice->data_details.ntc_259.gid2;
```

Since it can see that gid2 has a strange alignment.

However, there is now a subtle trap, where this will alignment trap on some arches:

```
extern void some_function(const ib_gid_t *arg);
some_function(&notice->data_details.ntc_259.gid2);
```

As the compiler in some_function() will assume that 'arg' has 8 byte alignment.

Generally speaking it is a dangerous idea to have structs off their defined alignment.

Within IB mads, GIDs are only aligned to at most 4 bytes, so a gid struct designed for use with mads should be set to packed or to align 4.